v10/cmd/lcc/etc/cc.1
.TH CC 1
.CT 1 prog_c
.SH NAME
cc, lcc \- C compilers
.SH SYNOPSIS
.B cc
[
.I option ...
]
.I file ...
.PP
.B lcc
[
.I option ...
]
.I file ...
.SH DESCRIPTION
.I Cc
compiles the classic C language;
.I lcc
compiles
.SM ANSI.
They are otherwise similar.
In the absence of options, any named source
.I files
are compiled into object files and then linked,
together with any named object
.I files,
into a single executable file named 
.FR a.out .
Compilation normally has four phases: preprocessing of
.B #
directives, compilation to assembly language, assembly,
and linking.
Suffixes of
.I file
names determine which phases they participate in:
.TP
.B .c
C source to be preprocessed and compiled.
Object code for this file is finally placed in a
correspondingly named
.L .o
file, except when exactly one file is being compiled and linked.
.TP
.B .i
C source to be compiled without preprocessing;
.B #
directives are ignored by
.I cc,
forbidden by
.I lcc.
.TP
.B .s
Assembler source to be assembled, producing a
.L .o
file.
.TP
.B .o
A preexisting object file to be linked.
.PP
Both compilers accept options of
.IR ld (1),
the most common of which are
.B -o
(to substitute a name for
.BR a.out )
and
.BR -l 
(to link from a library), and in addition
.TP
.B -c
Suppress the linking phase, producing
.B .o
files but no
.BR a.out .
.TP
.B -g
Produce additional symbol table information
for debuggers such as
.IR pi (9.1).
.TP
.B -O
Invoke an
object-code improver; superfluous in
.I lcc.
.TP
.B -w
Suppress warning diagnostics.
In
.I lcc,
.B #pragma ref
.I variable
supplies a dummy reference to suppress
an unused-variable diagnostic.
.TP
.B -p
Arrange for the compiler to produce code
which counts the number of times each routine is called;
also, if linking takes place, replace the standard startup
routine by one which arranges to gather profiling data
for later examination by
.IR  prof (1).
.TP
.B -pg
Like
.B -p
but for
.IR gprof 
instead of
.IR prof (1).
.TP
.B -S
Compile the named C programs, and leave the
assembler-language output in
.B .s
files.
.TP
.B -E
Run the preprocessor
on the named C programs, and send the result to the
standard output.
.TP
.B -C
Prevent the preprocessor from eliding comments.
.TP
.BI -D\*S name=def
.br
.ns
.TP
.BI -D \*Sname
Define the
.I name
to the preprocessor,
as if by
.LR #define .
If no definition is given, the name is defined as
.LR 1 .
.I Lcc
predefines a few symbols on most machines; option
.B -v
exposes them.
.TP
.BI -U \*Sname
Remove any initial definition of
.IR name .
.TP
.BI -I \*Sdir
.L #include
files whose names do not begin with 
.L /
are always
sought first in the directory 
of the
.I file
argument,
then in directories named in 
.B -I
options,
then in directories on a standard list.
.PP
These options are peculiar to
.IR cc :
.TP
.B -P
Run the preprocessor on each 
.L .c
file.
Produce no line numbers.
Place results in
.L .i
files.
.TP
.B -R
Cause
.IR as (1)
to make initialized variables shared and read-only.
.TP
.BI -B \*Sstring
Find substitute compiler passes in the files named
.I string
with the suffixes cpp, ccom and c2.
If 
.I string 
is empty, use a standard backup version.
.TP
.BR -t [ p012 ]
Find only the designated compiler passes in the
files whose names are constructed by a
.B -B
option.
In the absence of a
.B -B 
option, the
.I string
is taken to be
.FR /usr/c/ .
.PP
These options are peculiar to
.IR lcc :
.TP
.B -N
Do not search standard directories for include files.
Omit non-ANSI language extensions.
.TP
.B -A
Warn about calls to functions without prototypes.
.TP
.B -b
produce code that writes an expression-level profile into prof.out.
.IR bprint (1)
produces an annotated listing, and
.B -Wf-a
uses the profile to improve register assignments.
.TP
.BI \-d n
Generate jump tables for switches with density at least
.IR n ,
a floating-point constant between zero and one,
0.5 by default.
.TP 
.B \-P
Write declarations for all defined globals on standard error.
.TP
.B \-n
Produce code
that reports and aborts upon dereferencing a zero pointer.
.TP
.B \-M
Run only the preprocessor to generate 
.IR make (1)
dependencies on the standard output.
.TP
.B \-t
Produce code to print trace messages at function entry and exit.
.TP
.BI -Wp "opt"
Pass preprocessor option
.I opt
to the (Gnu) preprocessor.
For example,
.B -Wp-T
allows
.SM ANSI
trigraph sequences.
.HP
.BI -Wa  opt,
.BI -Wl  opt,
.BI -Wf  opt
.br
Pass option
.IR opt
to the assembler
.RI ( as (1)),
loader
.RI ( ld (1)),
or compiler proper.
.TP
.BI \-B str
Use the compiler
.IB str rcc
instead of the default version.
.I Str
usually ends with a slash.
.TP
.B \-v
Report compiler steps (and some version numbers) as
they are executed.
A second
.B \-v
causes steps to be reported but not executed.
.PP
.I Lcc
supports
.BI asm( string ).
The given string constant is copied to the generated
assembly language output with occurrences of
.BI % name
replaced by the address or register for identifier 
.I name 
if it is visible.
Otherwise,
.BI % name
is simply copied to the output.
Wide-character literals are treated as plain char literals;
ints and long ints are the same size,
as are doubles and long doubles.
.SH EXAMPLES
.TP
.L
lcc -N -I/usr/include/libc file.c 
Use local include files instead of
.SM ANSI
standard ones, which lack most functions of Section 2
of this manual, and often disagree (especially about
.BR const )
with those in Section 3.
See
.IR intro (3).
.SH FILES
.PP
Different machines use different file names, so this list is
only representative.
.I Lcc
option
.B -v
exposes the correct names.
.TF /usr/include/libc
.TP
.F a.out
linked output
.TP
.F /tmp/ctm*
temporary
.TP
.F /lib/cpp
preprocessor,
.IR cpp (8)
.TP
.F /usr/lib/gcc-cpp
.SM ANSI
preprocessor
.TP
.F /lib/ccom
.I cc
compiler proper
.TP
.F /lib/c2
optional optimizer for
.I cc
.TP
.F /bin/as
assembler,
.IR as (1)
.TP
.F /usr/lib/rcc
.I lcc
compiler proper
.TP
.F /lib/crt0.o
runtime startoff
.TP
.F /lib/mcrt0.o
startoff for profiling
.TP
.F /lib/libc.a
standard library, see
.IR intro (3)
.TP
.F /usr/include
directory for 
.I cc
.L #include
files
.TP
.F /usr/include/lcc
directory for 
.SM ANSI
standard 
.L #include
files
.TP
.F /usr/include/libc
directory for local
.I lcc
include files
.SH "SEE ALSO"
.IR lint (1), 
.IR ld (1), 
.IR strip (1),
.IR nm (1),
.IR prof (1), 
.IR bprint (1), 
.IR cin (1),
.IR adb (1), 
.IR pi (9.1),
.IR c++ (1)
.br
B. W. Kernighan and D. M. Ritchie,
.I The C Programming Language,
2nd Ed., Prentice-Hall, 1988
.SH BUGS
.I Cc
cannot handle the
.L -y
flag of
.I ld.
.br
.I Lcc
currently uses the pre-ANSI library.
.SH MACHINE DEPENDENCIES
.SS VAX
.B -pg
is unimplemented.
.br
.I Cc
and
.I lcc
use incompatible bit-field layouts and structure return conventions.
.SS MIPS
.I Lcc
does not implement
.B -p
or
.BR -pg ,
and its
.B -g
supports breakpoints but not the examination of variables.
.br
.I Cc
and
.I lcc
use incompatible bit-field layouts.
.SS Sun
.I Lcc
options
.B -Bdynamic
and
.B -Bstatic
give the binding strategy; see
.IR ld (1).
.br
.I Cc and
.I lcc
use incompatible bit-field layouts and structure return conventions.


v10/cmd/cc.c
#include <sys/types.h>
#include <stdio.h>
#include <ctype.h>
#include <signal.h>
#include <dir.h>

char	*cpp = "/lib/cpp";
char	*ccom = "/lib/ccom";
char	*c2 = "/lib/c2";
char	*as = "/bin/as";
char	*ld = "/bin/ld";
char	*crt0 = "/lib/crt0.o";
char	*instrcnt = "/lib/instrcnt";

char	tmp0[30];		/* big enough for /tmp/ctm%05.5d */
char	*tmp1, *tmp2, *tmp3, *tmp4, *tmp5;
char	*outfile;
char	*savestr(), *strspl(), *setsuf();
int	idexit();
char	**av, **clist, **llist, **plist;
int	cflag, eflag, gflag, oflag, pflag, sflag, wflag, Rflag, exflag, proflag;
int	vflag, Cflag;
char	*dflag;
int	exfail;
char	*chpass;
char	*npassname;
extern	int	optind;
extern	int	opterr;
extern	char	*optarg;
extern	int	optopt;

int	nc, nl, np, nxo, na;

#define	cunlink(s)	if (s) unlink(s)

main(argc, argv)
	char **argv;
{
	char *t;
	char *assource;
	int i, j, c;
	char errbuf[BUFSIZE];
	setbuf(stderr, errbuf);

	/* ld currently adds upto 5 args; 10 is room to spare */
	av = (char **)calloc(argc+10, sizeof (char **));
	clist = (char **)calloc(argc, sizeof (char **));
	llist = (char **)calloc(argc, sizeof (char **));
	plist = (char **)calloc(argc, sizeof (char **));
	opterr = 0;
	while (optind<argc) switch (c = getopt(argc, argv, "vsSo:RiOPgwEpPcD:I:U:C:t:B:l:d:")) {
	case 'i':
		Cflag++;
		continue;
	case 'v':
		vflag++;
		continue;
	case 'S':
		sflag++;
		cflag++;
		continue;
	case 'l':
		llist[nl++] = strspl("-l", optarg);
		continue;
	case 'o':
		outfile = optarg;
		switch (getsuf(outfile)) {

		case 'c':
		case 'o':
			error("-o would overwrite %s", outfile);
			exit(8);
		}
		continue;
	case 'R':
		Rflag++;
		continue;
	case 'O':
		oflag++;
		continue;
	case 'p':
		proflag++;
		continue;
	case 'g':
		gflag++;
		continue;
	case 'w':
		wflag++;
		continue;
	case 'E':
		exflag++;
	case 'P':
		pflag++;
		t = strspl("-", "x");
		t[1] = optopt;
		plist[np++] = t;
	case 'c':
		cflag++;
		continue;
	case 'D':
	case 'I':
	case 'U':
	case 'C':
		plist[np] = strspl("-X", optarg);
		plist[np++][1] = c;
		continue;
	case 't':
		if (chpass)
			error("-t overwrites earlier option", 0);
		chpass = optarg;
		if (chpass[0]==0)
			chpass = "012p";
		continue;
	case 'B':
		if (npassname)
			error("-B overwrites earlier option", 0);
		npassname = optarg;
		if (npassname[0]==0)
			npassname = "/usr/c/o";
		continue;
	case 'd':
		dflag = strspl("-d", optarg);
		continue;

	case '?':
	case 's':
		t = strspl("-", "x");
		t[1] = optopt;
		llist[nl++] = t;
		continue;

	case EOF:
		t = argv[optind];
		optind++;
		c = getsuf(t);
		if (c=='c' || c=='s' || c=='i' || exflag) {
			clist[nc++] = t;
			t = setsuf(t, 'o');
		}
		if (nodup(llist, t)) {
			llist[nl++] = t;
			if (getsuf(t)=='o')
				nxo++;
		}
	}
	if (gflag) {
		if (oflag)
			fprintf(stderr, "cc: warning: -g disables -O\n");
		oflag = 0;
	}
	if (npassname && chpass ==0)
		chpass = "012p";
	if (chpass && npassname==0)
		npassname = "/usr/new";
	if (chpass)
	for (t=chpass; *t; t++) {
		switch (*t) {

		case '0':
			ccom = strspl(npassname, "ccom");
			continue;
		case '2':
			c2 = strspl(npassname, "c2");
			continue;
		case 'p':
			cpp = strspl(npassname, "cpp");
			continue;
		}
	}
	if (proflag)
		crt0 = "/lib/mcrt0.o";
	if (nc==0)
		goto nocom;
	if (signal(SIGINT, SIG_IGN) != SIG_IGN)
		signal(SIGINT, idexit);
	if (signal(SIGTERM, SIG_IGN) != SIG_IGN)
		signal(SIGTERM, idexit);
	if (pflag==0)
		sprintf(tmp0, "/tmp/ctm%05.5d", getpid());
	tmp1 = strspl(tmp0, "1");
	tmp2 = strspl(tmp0, "2");
	tmp3 = strspl(tmp0, "3");
	if (pflag==0)
		tmp4 = strspl(tmp0, "4");
	if (oflag || Cflag)
		tmp5 = strspl(tmp0, "5");
	for (i=0; i<nc; i++) {
		int suffix = getsuf(clist[i]);
		if (nc > 1) {
			printf("%s:\n", clist[i]);
			fflush(stdout);
		}
		if (suffix == 's') {
			assource = clist[i];
			goto assemble;
		} else
			assource = tmp3;
		if (suffix == 'i')
			goto compile;
		if (pflag)
			tmp4 = setsuf(clist[i], 'i');
		av[0] = "cpp";
		av[1] = clist[i];
		av[2] = exflag ? "-" : tmp4;
		na = 3;
		for (j = 0; j < np; j++)
			av[na++] = plist[j];
		av[na++] = 0;
		switch (callsys(cpp, av)) {
			case 0:
				break;
#define CLASS 27
			case CLASS:
				if (callsys("/lib/cpre",av) == 0)
					break;
			default:
				exfail++;
				eflag++;
		}
		if (pflag || exfail) {
			cflag++;
			continue;
		}
compile:
		if (sflag)
			assource = tmp3 = setsuf(clist[i], 's');
		av[0] = "ccom";
		av[1] = suffix=='i'? clist[i]: tmp4;
		av[2] = (oflag || Cflag)? tmp5:tmp3;
		na = 3;
		if (proflag)
			av[na++] = "-Xp";
		if (gflag)
			av[na++] = "-g";
		if (wflag)
			av[na++] = "-w";
		if (oflag)
			av[na++] = "-XO";
		av[na] = 0;
		if (callsys(ccom, av)) {
			cflag++;
			eflag++;
			continue;
		}
		if (oflag || Cflag) {
			av[0] = Cflag? "instrcnt": "c2";
			av[1] = tmp5; av[2] = tmp3; av[3] = 0;
			if (callsys(Cflag? instrcnt: c2, av)) {
				unlink(tmp3);
				tmp3 = assource = tmp5;
			} else
				unlink(tmp5);
		}
		if (sflag)
			continue;
	assemble:
		cunlink(tmp1); cunlink(tmp2); cunlink(tmp4);
		av[0] = "as"; av[1] = "-o"; av[2] = setsuf(clist[i], 'o');
		na = 3;
		if (Rflag)
			av[na++] = "-R";
		if (dflag)
			av[na++] = dflag;
		av[na++] = assource;
		av[na] = 0;
		if (callsys(as, av)) {
			cflag++;
			eflag++;
			continue;
		}
	}
nocom:
	if (cflag==0 && nl!=0) {
		i = 0;
		av[0] = "ld"; av[1] = "-X"; av[2] = crt0; na = 3;
		if (outfile) {
			av[na++] = "-o";
			av[na++] = outfile;
		}
		while (i < nl)
			av[na++] = llist[i++];
		if (gflag)
			av[na++] = "-lsdb";
		av[na++] = "-lc";
		av[na++] = 0;
		eflag |= callsys(ld, av);
		if (nc==1 && nxo==1 && eflag==0)
			unlink(setsuf(clist[0], 'o'));
	}
	dexit();
}

idexit()
{

	eflag = 100;
	dexit();
}

dexit()
{

	if (!pflag) {
		cunlink(tmp1);
		cunlink(tmp2);
		if (sflag==0)
			cunlink(tmp3);
		cunlink(tmp4);
		cunlink(tmp5);
	}
	exit(eflag);
}

error(s, x)
	char *s, *x;
{
	FILE *diag = exflag ? stderr : stdout;

	fprintf(diag, "cc: ");
	fprintf(diag, s, x);
	putc('\n', diag);
	fflush(diag);
	exfail++;
	cflag++;
	eflag++;
}

getsuf(as)
char as[];
{
	register int c;
	register char *s;
	register int t;

	s = as;
	c = 0;
	while (t = *s++)
		if (t=='/')
			c = 0;
		else
			c++;
	s -= 3;
	if (c <= DIRSIZ && c > 2 && *s++ == '.')
		return (*s);
	return (0);
}

char *
setsuf(as, ch)
	char *as;
{
	register char *s, *s1;

	s = s1 = savestr(as);
	while (*s)
		if (*s++ == '/')
			s1 = s;
	s[-1] = ch;
	return (s1);
}

callsys(f, v)
	char *f, **v;
{
	int t, status;
	register char **vp;	/* god */

	if(vflag) {	/* god & mjm */
		vp = v;
		fprintf(stderr,"+ ");
		while (*vp)
			fprintf(stderr,"%s ",*vp++);
		fprintf(stderr, "\n");
		fflush(stderr);
	}


	t = fork();
	if (t == -1) {
		printf("No more processes\n");
		return (100);
	}
	if (t == 0) {
		execv(f, v);
		printf("Can't find %s\n", f);
		fflush(stdout);
		_exit(100);
	}
	while (t != wait(&status))
		;
	if ((t=(status&0377)) != 0 && t!=14) {
		if (t!=2) {
			printf("Fatal error in %s\n", f);
			eflag = 8;
		}
		dexit();
	}
	return ((status>>8) & 0377);
}

nodup(l, os)
	char **l, *os;
{
	register char *t, *s;
	register int c;

	s = os;
	if (getsuf(s) != 'o')
		return (1);
	while (t = *l++) {
		while (c = *s++)
			if (c != *t++)
				break;
		if (*t==0 && c==0)
			return (0);
		s = os;
	}
	return (1);
}

#define	NSAVETAB	1024
char	*savetab;
int	saveleft;

char *
savestr(cp)
	register char *cp;
{
	register int len;

	len = strlen(cp) + 1;
	if (len > saveleft) {
		saveleft = NSAVETAB;
		if (len > saveleft)
			saveleft = len;
		savetab = (char *)malloc(saveleft);
		if (savetab == 0) {
			fprintf(stderr, "ran out of memory (savestr)\n");
			exit(1);
		}
	}
	strncpy(savetab, cp, len);
	cp = savetab;
	savetab += len;
	saveleft -= len;
	return (cp);
}

char *
strspl(left, right)
	char *left, *right;
{
	char buf[BUFSIZ];

	strcpy(buf, left);
	strcat(buf, right);
	return (savestr(buf));
}
